### PR TITLE
[v1.1][ISSUE-202][Code Refine for Performance Test]

### PR DESCRIPTION
**Pull request by XinyaoWa**
Model Adapter:
- Add/update finetuner and distiller realted configuration files to meet performance test requirements
- In task.py, do not create distiller/adapter/finetuner to match strategy, which can save useless definition
-  in train.py, disable record parameters in tensorboard, which can save a lot of time
- bug fix
     - tensorboard dir will be created in default even user don't need it, fix it in main.py
     - in train.py, fix DDP not sync according to Yu's PR
   

Common Lib
- In CV dataloader, find evaluation data will be split in distribution mode, which is not proper in some cases. Add an option "ddp_eval_nosplit" to make a complete evaluation.
- Bug fix
      - For Earlystop in utils.py, disable tolerance when set target metric

scripts/start_e2eaiok_docker.py
- Add no_push option, can be used if have local docker image installed.

### CODE REVIEW

**Comment by XinyaoWa** [comment]
@Peach-He Please have a review for common lib part
@zhouyu5 Please have a review for Model Adapter part

**Review by Peach-He — APPROVED** [review]

File: e2eAIOK/common/trainer/utils/utils.py [hunk 1]
```diff
@@ -189,23 +189,24 @@ def __init__(self, tolerance_epoch = 5, delta=0, is_max = False, metric_threshol
...
-        ############## relative level #################
-        if (self._is_max and (validation_metric < optimal_metric - self._delta)) \
-                or ((not self._is_max) and (validation_metric > optimal_metric + self._delta)): # less optimal
-            self._counter += 1
-        else: # more optimal
-            logging.info("Reset earlystop counter")
-            self._counter = 0
-        if self._counter >= self._tolerance_epoch:
-            self.early_stop = True
+        else:
```
Comment (xuechendi) [inline e2eAIOK/common/trainer/utils/utils.py:200]: @XinyaoWa , when 'else' path will be called? This fix is for which bullet in your description?

File: e2eAIOK/common/trainer/data/data_builder_cv.py [hunk 2]
```diff
@@ -21,20 +21,26 @@ def get_dataloader(self):
...
         if ext_dist.my_size > 1:
             num_tasks = ext_dist.dist.get_world_size()
             global_rank = ext_dist.dist.get_rank()
             sampler_train = torch.utils.data.DistributedSampler(
                 self.dataset_train, num_replicas=num_tasks, rank=global_rank, shuffle=True, drop_last= drop_last)
+        else:
+            sampler_train = torch.utils.data.RandomSampler(self.dataset_train)
+
+        ## For evaluation
+        if ext_dist.my_size > 1 and not ("ddp_eval_nosplit" in self.cfg and self.cfg.ddp_eval_nosplit): # ddp_eval_nosplit flag means do not split evaluation dataset in distribution mode
```
Comment (xuechendi) [inline e2eAIOK/common/trainer/data/data_builder_cv.py:34]: How to decide when we should use or not use 'ddp_eval_nosplit'

File: e2eAIOK/ModelAdapter/training/train.py [hunk 3]
```diff
@@ -112,11 +112,18 @@ def _dist_wrapper(self):
...
-            if self.is_transferrable: # dispatch method to model.module
-                self.model.get_backbone = MethodType(lambda obj: obj.module.get_backbone() ,self.model) # obj stands for model
-                self.model.get_training_metrics = MethodType(lambda obj,*args: obj.module.get_training_metrics(*args) ,self.model) #  obj stands for model
+            if self.is_transferrable:
+                self.model.backbone = ext_dist.DDP(self.model.backbone)
+                self.model.backbone.loss = MethodType(lambda obj,*args: obj.module.loss(*args) ,self.model.backbone) #  obj stands for model, both original model and transferrable model
+                if self.model.distiller is not None and not self.model.distiller.is_frozen:
+                    self.model.distiller = ext_dist.DDP(self.model.distiller) # if distiller require_grad=False, may fail in this step, just skip
+                if self.model.adapter is not None:
+                    self.mode.adapter = ext_dist.DDP(self.mode.adapter)
```
Comment (xuechendi) [inline e2eAIOK/ModelAdapter/training/train.py:121]: what will happen if the configuration not either met line 118 condition or line 120 condition? Does that mean it will directly use no DDP?

File: e2eAIOK/common/trainer/utils/utils.py (hunk 1 above)
Comment (XinyaoWa) [reply e2eAIOK/common/trainer/utils/utils.py:200]: If not set metric_threshold, else will be called, in this case, training will stop when acc not increase for a while. 

This fixs "For Earlystop in utils.py, disable tolerance when set target metric". Before fix, even metric_threshold is set, "stop when acc not increase for a while" will also be applied, which is harmful when we want to meet some target score; thus fix it.

File: e2eAIOK/common/trainer/data/data_builder_cv.py (hunk 2 above)
Comment (XinyaoWa) [reply e2eAIOK/common/trainer/data/data_builder_cv.py:34]: Depends on your use case. 
If evaluation tooks too long time and the metric is simple, can split it and then merge.
But some complex metric is not proper to be merged, such as if they depends on the whole data. Or evaluation is very quick, we can calculate once to skip spliting and merging.

File: e2eAIOK/ModelAdapter/training/train.py [hunk 4]
```diff
@@ -112,11 +112,18 @@ def _dist_wrapper(self):
...
+            else:
+                self.model = ext_dist.DDP(self.model)
+                self.model.loss = MethodType(lambda obj,*args: obj.module.loss(*args) ,self.model) #  obj stands for model, both original model and transferrable model
+                #     self.model.get_backbone = MethodType(lambda obj: obj.module.get_backbone() ,self.model) # obj stands for model
+                #     self.model.get_training_metrics = MethodType(lambda obj,*args: obj.module.get_training_metrics(*args) ,self.model) #  obj stands for model
```
Comment (zhouyu5) [inline e2eAIOK/ModelAdapter/training/train.py:126]: Here you may not comment these two lines, because your model is wrapped by DDP, directly take the attribute will report a error, please test this use case: when you use DDP and your model is not transferrable

File: e2eAIOK/ModelAdapter/training/train.py (hunk 3 above)
Comment (XinyaoWa) [reply e2eAIOK/ModelAdapter/training/train.py:121]: 118 and 120 is to warp the submodule of transferrable mode.
Here is the full code, if it's not a transferrable model, it will use normal DDP. 
image

File: e2eAIOK/common/trainer/utils/utils.py (hunk 1 above)
Comment (xuechendi) [reply e2eAIOK/common/trainer/utils/utils.py:200]: I see, thanks for the explanation.

File: e2eAIOK/common/trainer/data/data_builder_cv.py (hunk 2 above)
Comment (xuechendi) [reply e2eAIOK/common/trainer/data/data_builder_cv.py:34]: I got that, my concern is that 'ddp_eval_nosplit' is an advanced configuration and required user already did recursive tests and they are finding current result is not bad. Which will be a bad user experience.
My suggestion here is for 'known' dataset, can we give default configuration on 'dpp_eval_nosplit' and only when this is 'unknown'/'new' dataset,  'ddp_eval_nosplit' is required user's configuration.

File: e2eAIOK/ModelAdapter/training/train.py (hunk 3 above)
Comment (xuechendi) [reply e2eAIOK/ModelAdapter/training/train.py:121]: which line is for normal DPP? if configuration shows it is 'is_transferrable' but not fall into line 118 and line 120, it also won't fall into else. So it will leave ext_dist.my_size > 1 condition.

File: e2eAIOK/ModelAdapter/training/train.py (hunk 4 above)
Comment (XinyaoWa) [reply e2eAIOK/ModelAdapter/training/train.py:126]: backbone.loss is applied in task.py. 125 and 126 is not needed when it's not a transferrable mode.

File: e2eAIOK/ModelAdapter/training/train.py (hunk 3 above)
Comment (XinyaoWa) [reply e2eAIOK/ModelAdapter/training/train.py:121]: The components of transferrable model are: backbone, finetuner, distiller, adapter.
line 116 deals with backbong.
Finetuner's function is to apply parmeter, no need for training.
line 118 and 120 will deal distiller and adapter.
No other component will appear in transferrable model.

File: e2eAIOK/common/trainer/data/data_builder_cv.py (hunk 2 above)
Comment (XinyaoWa) [reply e2eAIOK/common/trainer/data/data_builder_cv.py:34]: We have default option for "known" dataset, all DENAS related cases will split evaluation data; all Model Adapter related cases will not split, this is defined in "e2eAIOK/ModelAdapter/default_ma.conf".

**Review by zhouyu5 — APPROVED** [review]

**Comment by XinyaoWa** [comment]
@xuechendi  code is ready to merge.

File: e2eAIOK/ModelAdapter/training/train.py (hunk 3 above)
Comment (xuechendi) [reply e2eAIOK/ModelAdapter/training/train.py:121]: will there be any case falls to self.model.distiller is not None and self.model.distiller.is_frozen ?
Accoding to SDL guideline, usually we should put some else raise to cover any unexpected scenario.
I felt currently you only covered expected scenarios, but will there be any user configuration that goes to condition not covered above?

Comment (XinyaoWa) [reply e2eAIOK/ModelAdapter/training/train.py:121]: "self.model.distiller is not None and self.model.distiller.is_frozen" In this case, as it's parameter is frozen, distiller will not be trained and don't need to apply DDP.
"usually we should put some else raise to cover any unexpected scenario." - Yes, but in this case, I don't find any unexpected scenarios, welcome for any other thoughts~

File: e2eAIOK/ModelAdapter/default_ma.conf [hunk 5]
```diff
@@ -14,6 +14,7 @@ experiment:
     strategy: ""
 
 # Dataset
+ddp_eval_nosplit: True
```
[the remark this answers is not in the document]
Comment (XinyaoWa) [reply e2eAIOK/ModelAdapter/default_ma.conf:17]: For other dataset, if do not define ddp_eval_nosplit in dataloader, this parameter will just be ignored.

**Review by xuechendi — APPROVED** [review]
approved
